Implement URLPatternInit support for URLPatternAPI.

jelee53 opened the pull request:
#### 1e937703de6af995ef2cd93867292c9da1d3b86c
<pre>
Implement URLPatternInit support for URLPatternAPI.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281711">https://bugs.webkit.org/show_bug.cgi?id=281711</a>
<a href="https://rdar.apple.com/138169179">rdar://138169179</a>

Reviewed by Chris Dumez and Sihui Liu.

URLPattern API spec for processing a URLPatternInit: <a href="https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit.">https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit.</a>

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::escapePatternStringForCharacters):
(WebCore::escapePatternString):
(WebCore::processBaseURLString):
(WebCore::processInit):
(WebCore::URLPattern::create):
(WebCore::URLPattern::URLPattern):
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp: Added.
(WebCore::isInvalidIPv6HostCodePoint):
(WebCore::isAbsolutePathname):
(WebCore::canonicalizeProtocol):
(WebCore::canonicalizeUsername):
(WebCore::canonicalizePassword):
(WebCore::canonicalizeHost):
(WebCore::canonicalizeIPv6Host):
(WebCore::canonicalizePort):
(WebCore::canonicalizeOpaquePath):
(WebCore::canonicalizePath):
(WebCore::canonicalizeSearch):
(WebCore::canonicalizeHash):
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.h: Copied from Source/WebCore/Modules/url-pattern/URLPattern.cpp.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/285640@main">https://commits.webkit.org/285640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c4d107563bc174fb9d7b881ac9b374925717796

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73301 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52730 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77532 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24539 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/513 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57589 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16060 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47625 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63072 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38008 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44264 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20552 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22868 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66121 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20905 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79183 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/147 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66020 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/758 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63082 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65301 "Found 1 new API test failure: /TestWebKit:WebKit.GetInjectedBundleInitializationUserDataCallback (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9148 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7308 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11297 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/580 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3317 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/610 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/594 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+{
+    URLPatternInit result;
+
+    if (!argProtocol.isEmpty())
```
All these `isEmpty()` checks seem to achieve nothing. `result.protocol = argProtocol;` wouldn't be harmful or have any side effects that I can see. First of all, the spec says to check for `null`, not `empty`. So in theory, this should have been a `isNull()` check. However, since `result.argProtocol` is initially null, then just `result.protocol = argProtocol;` would be a no-op if `argProtocol` was null. therefore, I don't think the `if` checks are needed.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+}
+
+static URLPatternInit processInit(URLPatternInit&& init, BaseURLStringType type, String&& argProtocol, String&& argUsername,
+    String&& argPassword, String&& argHostname, String&& argPort, String&& argPathname, String&& argSearch, String&& argHash)
```
I think we can omit the `arg` prefix to all parameters. They don't seem useful and the WebKit coding style says no abbreviations.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+
+    URL baseURL;
+
+    if  (!(init.baseURL.isEmpty())) {
```
Too many parentheses here and also, the check should be isNull(), should be:
```
if  (!init.baseURL.isNull()) {
```

The spec says `If init["baseURL"] exists`. It could exist and be the empty string. If it's missing the String will be null, not empty.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+    URL baseURL;
+
+    if  (!(init.baseURL.isEmpty())) {
+
```
Please omit this blank line.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+
+        if (!baseURL.isValid() || baseURL.isEmpty()) {
+            // FIXME // Throw Type Error Exception // Log error for now
+            WTFLogAlways("baseURL is not valid");
```
We cannot land a WTFLogAlways.

cdumez replied:
The spec says to throw an exception if baseURL fails parsing. Seems this function should return an `ExceptionOr<URLPatternInit>` and we should return an Exception here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+            WTFLogAlways("baseURL is not valid");
+        }
+
+        if (init.protocol.isEmpty())
```
Please double check all usage of isEmpty() vs isNull() in this PR.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+    return result.toString();
+}
+
+static String processBaseURLString(String&& input, BaseURLStringType type)
```
Please add a comment with a link to the spec:
```
// https://urlpattern.spec.whatwg.org/#process-a-base-url-string
```

cdumez replied:
`input` should be a StringView, then the call sites do not have to call `toString()`.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
 
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(URLPattern);
 
+static String escapeStringHelper(StringView input)
```
Please add a comment with a link to the spec:
```
// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
 
+static String escapeStringHelper(StringView input)
+{
+    StringBuilder result;
```
The spec says to do:
```
ASSERT(input.containsOnlyASCII());
```

cdumez replied:
Would also be nice to call `result.reserveCapacity(input.size());` since this will be the common case. It will avoid a lot of memory reallocations.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+    StringBuilder result;
+    char escapeCharacters[] = "\\+*?:{}()";
+
+    for (auto character : input.span8()) {
```
using `span8()` here is incorrect. The String could be backed by 16 bits characters. This should be:
`for (UChar character : input) {

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+static String escapeStringHelper(StringView input)
+{
+    StringBuilder result;
+    char escapeCharacters[] = "\\+*?:{}()";
```
We don't use old-style C arrays in new code anymore for safety reasons (since they don't bounds checks). This should be a `std::array`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+    char escapeCharacters[] = "\\+*?:{}()";
+
+    for (auto character : input.span8()) {
+        if (strchr(escapeCharacters, character))
```
Should probably switch to using `std::find()` once `escapeCharacters` becomes a `std::array`.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,205 @@
+    URLPatternInit init;
+
+    if (!argInput.has_value()) {
+        // FIXME // TODO Throw Error
```
Our format is:
```
// FIXME: Throw an exception.
```

cdumez replied:
`FIXME:` + sentence ending with a period. Per WebKit coding style.

cdumez replied:
But also, where in the spec does it say to throw? Honestly, since the spec uses an optional union with a default value, I'd expect a missing parameter to be interpreted as a default initialized first type of the union, so same as `std::holds_alternative<String>(argInput.value())` with a default-constructed String.

jelee53 replied:
It doesn't say anywhere in the spec. I was just adding it for additional safety but you're right that it would eventually be evaluated as a null string. Removed!

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -36,6 +36,7 @@ namespace WebCore {
 struct URLPatternInit;
 struct URLPatternOptions;
 struct URLPatternResult;
+enum class BaseURLStringType :uint8_t { Pattern, URL };
```
missing space after `:`

cdumez reviewed (commented):
I haven't done a full review yet but I suspect some of the comments I made will apply to the rest of the PR too. I'll wait for the next iteration to review more.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+template<typename CharacterType>
+static String escapeStringForCharacters(std::span<const CharacterType> characters)
+{
+    std::array<UChar, 9> escapeCharacters = { '+', '*', '?', ':', '(', ')', UChar(0x005C /* U+005C is '\' */), UChar(0x007B /* U+007B is '{' */), UChar(0x007D /* U+007D is '}' */) };
```
I think this array can be `static constexpr`. Also, I think we can use `LChar` type for storage instead of `UChar` given that all characters are latin1. finally, I think we use can '{' instead of `UChar(0x007B)` (same for other characters).

jelee53 replied:
I can't build if I do '\\'. Also style-checker complains about brackets, too.

cdumez replied:
Use `'\\'` to fix the build. You can also likely omit the `= `

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStringType type, String&& protocol, String&& username,
+    String&& password, String&& hostname, String&& port, String&& pathname, String&& search, String&& hash)
+{
+    UNUSED_PARAM(type);
```
This is used.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    URLPatternInit result;
+    URL baseURL;
+
+    if  (!init.baseURL.isNull()) {
```
We're missing steps 2 through 9 in the spec (which don't seem like a lot of code) so we should probably include them. If it's not trivial for some reason, I would at least add a FIXME comment.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+        baseURL = URL(init.baseURL);
+
+        if (!baseURL.isValid() || baseURL.isEmpty())
+            return Exception { ExceptionCode::TypeError };
```
Please specify a second parameter, with the exception. E.g.:
```
return Exception { ExceptionCode::TypeError, "Invalid baseURL"_s };
```
It is bad practice to throw an exception without letting the developer know what they did wrong.

rniwa commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+template<typename CharacterType>
+static String escapeStringForCharacters(std::span<const CharacterType> characters)
+{
+    std::array<UChar, 9> escapeCharacters = { '+', '*', '?', ':', '(', ')', UChar(0x005C /* U+005C is '\' */), UChar(0x007B /* U+007B is '{' */), UChar(0x007D /* U+007D is '}' */) };
```
I'm curious. Why can't we use '\\', '{', and '}'?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+
+        if (init.protocol.isNull()
+            && init.hostname.isNull()) {
+            String baseHost = (!baseURL.host().isNull()) ? baseURL.host().toString() : ""_s;
```
`(!baseURL.host().isNull())` -> `!baseURL.host().isNull()`

cdumez replied:
`""_s` -> `emptyString()`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            && init.hostname.isNull()
+            && init.port.isNull()) {
+            auto port = baseURL.port();
+            result.port = port ? String::number(*port) : ""_s;
```
`""_s` -> `emptyString()`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            && init.hostname.isNull()
+            && init.port.isNull()
+            && init.pathname.isNull()) {
+            String serializedBaseURL = baseURL.path().toString();
```
processBaseURLString() takes in a `StringView` so we don't need to call `toString()` here. I also don't think we need a local variable here and I would just do:
```
result.pathname = processBaseURLString(baseURL.path(), type);
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            && init.pathname.isNull()
+            && init.search.isNull()) {
+            String urlQuery = baseURL.hasQuery() ? baseURL.query().toString() : ""_s;
+            result.search = processBaseURLString(urlQuery, type);
```
Same comment about not needing a String, we could do:
```
result.search = processBaseURLString(baseURL.hasQuery() ? baseURL.query() : StringView { emptyString() }, type);
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            && init.pathname.isNull()
+            && init.search.isNull()
+            && init.hash.isNull()) {
+            String urlFragment = baseURL.hasFragmentIdentifier() ? baseURL.fragmentIdentifier().toString() : ""_s;
```
Same comment as above.

rniwa commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (value.isEmpty())
+        return value;
+
+    String strippedValue = value.endsWith(":"_s) ? value.substring(0, value.length()-1) : value;
```
Nit: Need a space around `-`.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (value.isEmpty())
+        return value;
+
+    String strippedValue = value.endsWith(":"_s) ? value.substring(0, value.length()-1) : value;
```
`value.endsWith(':')` would be a bit more efficient.

cdumez replied:
`value.substring(0, value.length()-1)` -> `value.left(value.length() - 1)`

cdumez replied:
Could use `auto`

rniwa commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (valueType == BaseURLStringType::Pattern)
+        return value;
+
+    URL dummyURL("https://www.webkit.org"_s);
```
Hm... this will create a dummy URL to canonicalize each component. That seems awfully inefficient.
Not that perf matters here but perhaps these functions should form a class with a URL as an instance variable.

rniwa replied:
On the other hand, spec does say to create dummy URL in each canonicalization step so maybe the current code is okay.

cdumez replied:
I think this is OK at least for now.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    URL dummyURL(makeString(strippedValue, "://webkit.test"_s));
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
```
Needs an exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    }
+
+    if (!init.protocol.isNull()) {
+        ExceptionOr<String> protocolResult = canonicalizeProtocol(WTFMove(init.protocol), type);
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+        result.password = canonicalizePassword(WTFMove(init.password), type);
+
+    if (!init.hostname.isNull()) {
+        ExceptionOr<String> hostResult = canonicalizeHost(WTFMove(init.protocol), type);
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    dummyURL.setHost(value);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+
+static bool isInvalidIPv6HostCodePoint(auto codepoint)
+{
+    std::array<UChar32, 3> invalidSpecialCodepoints = { '[', ']', ':' };
```
`static constexpr`? Also, can we use `LChar` for storage?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (input.isEmpty())
+        return false;
+
+    if (input.startsWith("/"_s))
```
`'/'` would be a bit more efficient.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+
+    for (auto codepoint : StringView(value).codePoints()) {
+        if (isInvalidIPv6HostCodePoint(codepoint))
+            return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    dummyURL.setPort(parseInteger<uint16_t>(portValue));
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
+
+    int portAsInteger = static_cast<int>(dummyURL.port().value());
```
Why the past to int here? Can't we just return `String::number(dummyURL.port())` below?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    dummyURL.setPath(value);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+        || protocolValue == "http"_s || protocolValue == "https"_s
+        || protocolValue == "ws"_s || protocolValue == "wss"_s) {
+
+        bool hasLeadingSlash = pathnameValue.startsWith("/"_s);
```
`'/'` would be more efficient.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+        dummyURL.setPath(maybeAddSlashPrefix);
+
+        if (!dummyURL.isValid())
+            return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (value.isEmpty())
+        return value;
+
+    String strippedValue = value.startsWith("?"_s) ? value.substring(1) : value;
```
`'?'`

cdumez replied:
and use `auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    dummyURL.setQuery(strippedValue);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    if (value.isEmpty())
+        return value;
+
+    String strippedValue = value.startsWith("#"_s) ? value.substring(1) : value;
```
Use `auto` and `'#'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+    dummyURL.setFragmentIdentifier(strippedValue);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
```
Please provide exception message.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+        if (!dummyURL.isValid())
+            return Exception { ExceptionCode::TypeError };
+
+        String result = dummyURL.path().toString();
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+        bool hasLeadingSlash = pathnameValue.startsWith("/"_s);
+        String modifiedValue = hasLeadingSlash ? ""_s : "/-"_s;
+
+        String maybeAddSlashPrefix = makeString(modifiedValue, pathnameValue);
```
`auto

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+        || protocolValue == "ws"_s || protocolValue == "wss"_s) {
+
+        bool hasLeadingSlash = pathnameValue.startsWith("/"_s);
+        String modifiedValue = hasLeadingSlash ? ""_s : "/-"_s;
```
`auto`. Note that this will resolve as an ASCIILiteral instead of constructing a String, which will be more efficient here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,258 @@
+        if (isInvalidIPv6HostCodePoint(codepoint))
+            return Exception { ExceptionCode::TypeError };
+
+        result.append(codepoint);
```
I think we're missing a call to `toASCIILower()` here per the spec.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    }
+
+    if (!init.port.isNull()) {
+        ExceptionOr<String> portResult = canonicalizePort(WTFMove(init.port), WTFMove(init.protocol), type);
```
`auto`

cdumez replied:
We need to be careful with all these `WTFMove()` calls, at line 176, you already move `init.protocol` so I think we're doing a potential use-after-move here. I wouldn't be surprised if we had the same issue in other places. My advice would be to stop using `WTMove()` on `init` values.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+
+        if (!baseURL.isNull() && baseURL.hasOpaquePath() && !isAbsolutePathname(result.pathname, type)) {
+            String baseURLPath = processBaseURLString(baseURL.path(), type);
+            size_t slashIndex = baseURLPath.reverseFind("/"_s);
```
`'/'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            String baseURLPath = processBaseURLString(baseURL.path(), type);
+            size_t slashIndex = baseURLPath.reverseFind("/"_s);
+
+            if (slashIndex != WTF::notFound)
```
`WTF::` is not needed.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            size_t slashIndex = baseURLPath.reverseFind("/"_s);
+
+            if (slashIndex != WTF::notFound)
+                result.pathname = makeString(baseURLPath.substring(0, slashIndex + 1), result.pathname);
```
use `StringView::left()`:
```
result.pathname = makeString(StringView { baseURLPath }.left(slashIndex + 1), result.pathname);
```

cdumez replied:
`left()` is just nicer syntax for this use case. Using StringView() avoid allocating a String for no reason.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+            if (slashIndex != WTF::notFound)
+                result.pathname = makeString(baseURLPath.substring(0, slashIndex + 1), result.pathname);
+
+            ExceptionOr<String> pathResult = canonicalizePath(WTFMove(result.pathname), baseURL.protocol(), type);
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    }
+
+    if (!init.search.isNull()) {
+        ExceptionOr<String> queryResult = canonicalizeSearch(WTFMove(init.search), type);
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    }
+
+    if (!init.hash.isNull()) {
+        ExceptionOr<String> fragmentResult = canonicalizeHash(WTFMove(init.hash), type);
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
 }
 
-Ref<URLPattern> URLPattern::create(std::optional<URLPatternInput>&&, URLPatternOptions&&)
+ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&& argInput, URLPatternOptions&&)
```
`input`. `argInput` is not a good name as we don't like abbreviations in WebKit.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
-    return adoptRef(*new URLPattern());
+    URLPatternInit init;
+
+    if (std::holds_alternative<String>(argInput.value())) {
```
This uses `argInput.value()` without checking that `argInput` is set (could be nullopt). This would crash.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    } else if (std::holds_alternative<URLPatternInit>(argInput.value()))
+        init = std::get<URLPatternInit>(argInput.value());
+
+    ExceptionOr<URLPatternInit> maybeProcessedInit = processInit(WTFMove(init), BaseURLStringType::Pattern, ""_s , String(), String(), String(), String(), String(), String(), String());
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
+    return adoptRef(*new URLPattern(WTFMove(processedInit.protocol), WTFMove(processedInit.username), WTFMove(processedInit.password), WTFMove(processedInit.port), WTFMove(processedInit.search), WTFMove(processedInit.hash)));
 }
 
+URLPattern::URLPattern(String&& argProtocol, String&& argUsername, String&& argPassword, String&& argPathname, String&& argSearch, String&& argHash) : m_protocol(argProtocol), m_username(argUsername), m_password(argPassword), m_port(argPathname), m_pathname(argSearch), m_search(argSearch), m_hash(argHash) { };
```
please don't use `arg` in parameter names.

cdumez replied:
Please use one line for each data member initialization, like we do elsewhere.

jelee53 replied:
Oops my bad! Forgot to fix these "arg" parameters from the last PR! Thanks Chris!

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -62,6 +63,7 @@ class URLPattern final : public RefCounted<URLPattern> {
     bool hasRegExpGroups() const { return m_hasRegExpGroups; }
 
 private:
+    URLPattern(String&& argProtocol, String&& argUsername, String&& argPassword, String&& argPathname, String&& argSearch, String&& argHash);
```
Please don't use `arg` in parameter names.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,239 @@
 {
     UNUSED_PARAM(baseURL);
 
     return adoptRef(*new URLPattern());
```
Seems be need a FIXME comment here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+template<typename CharacterType>
+static String escapeStringForCharacters(std::span<const CharacterType> characters)
+{
+    static constexpr std::array<LChar, 9> escapeCharacters = { '+', '*', '?', ':', '(', ')', '\\', '{', '}' }; // NOLINT
```
Would this build or does it really have to be `LChar`?
```
static constexpr std::array escapeCharacters = { '+', '*', '?', ':', '(', ')', '\\', '{', '}' };
```

jelee53 replied:
It does build but I thought we explicitly wanted it to be LChar? I'll change it back to that then for both escapeStringForCharacters and isValidIpv6Host

szewai replied:
Why does it need NOLINT?

jelee53 replied:
I was having trouble with the linter complaining that the brackets needed spaces.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    StringBuilder result;
+    result.reserveCapacity(characters.size());
+
+    for (auto& character : characters) {
```
`auto` instead of `auto&` since this is a LChar/UChar.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+}
+
+// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+static String escapeString(StringView input)
```
Maybe we call this `escapePatternString` to match the spec more closely and so that the name is not so generic?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    String&& password, String&& hostname, String&& port, String&& pathname, String&& search, String&& hash)
+{
+    URLPatternInit result;
+    result.protocol = protocol;
```
Can't we `WTFMove()` here and for all the non-init parameters below? Are they used later in the function?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+        result.pathname = init.pathname;
+
+        if (!baseURL.isNull() && baseURL.hasOpaquePath() && !isAbsolutePathname(result.pathname, type)) {
+            String baseURLPath = processBaseURLString(baseURL.path(), type);
```
`auto`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
 {
     UNUSED_PARAM(baseURL);
 
+    // FIXME: Implement URLPattern::create
```
`// FIXME: Implement URLPattern::create().`

Comments in WebKit need to end with a period.

szewai replied:
It might be better return an Exception instead of an empty object so the result is more observable:
`return Exception { ExceptionCode::NotSupportedError, "Not implemented"_s };`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
-    return adoptRef(*new URLPattern());
+    URLPatternInit init;
+
+    if (std::holds_alternative<String>(input.value()) && !std::get<String>(input.value()).isNull()) {
```
Still calling `input.value()` without checking if the optional is actually set?

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternInit.h:
```diff
@@ -25,8 +25,6 @@
 
 #pragma once
 
-#include <wtf/text/WTFString.h>
```
Why remove this? Seems useful given the structure below.

jelee53 replied:
Seemed to build without it so I assumed that somehow the build system was able to link String class to this header file. I'll add it back.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
 }
 
-URLPattern::URLPattern() = default;
+URLPattern::URLPattern(String&& protocol, String&& username, String&& password, String&& hostname, String&& port, String&& pathname, String&& search, String&& hash)
```
This constructor should likely just take a `URLPatternInit&&` to avoid so many parameters.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    auto processedInit = maybeProcessedInit.releaseReturnValue();
+
+    // FIXME: Implement compiling the component for each URLPattern member field to replace this current placeholder code which reroutes processed URLPatternInit object to URLPattern result.
+    return adoptRef(*new URLPattern(WTFMove(processedInit.protocol), WTFMove(processedInit.username), WTFMove(processedInit.password), WTFMove(processedInit.hostname), WTFMove(processedInit.port), WTFMove(processedInit.pathname), WTFMove(processedInit.search), WTFMove(processedInit.hash)));
```
I would just pass `WTFMove(WTFMove(processedInit)` here.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -36,14 +36,15 @@ namespace WebCore {
 struct URLPatternInit;
 struct URLPatternOptions;
 struct URLPatternResult;
+enum class BaseURLStringType : uint8_t { Pattern, URL };
```
`: bool`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+
+static bool isInvalidIPv6HostCodePoint(auto codepoint)
+{
+    static constexpr std::array<LChar, 3>  invalidSpecialCodepoints =  { '[', ']', ':' };
```
to many spaces before `invalidSpecialCodepoints`. the `= ` is not necessary.
Also, do we really need `<LChar, 3>`?

szewai replied:
These are **valid** code points in IPV6 (the check below returns true if the parameter does not belong to them).
Also, it may be faster to check if codepoint is valid (matching any of the conditions) than is invalid (matching all conditions).

cdumez replied:
I don't understand Sihui's comment (not sure if it's related to mine).

szewai replied:
Not quite related but for the same line: `invalidSpecialCodepoints` => `validSpecialCodepoints`

jelee53 replied:
oh yup I get it! Good catch Sihui! 

I'll rename the array to be validSpecialCodepoints! Because they are valid haha.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+    if (input.isEmpty())
+        return false;
+
+    if (input.startsWith('/'))
```
can just use `input[0] == '/'` given the `isEmpty()` check above.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    URL dummyURL("https://www.webkit.org"_s);
```
`"https://www.webkit.org"_s` is used in several places. Ideally we'd have a global constant for it:
```
static constexpr auto dummyURLCharacters { "https://www.webkit.org"_s };
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+
+    URL dummyURL("https://www.webkit.org"_s);
+
+    if (protocolValue.has_value())
```
Could just be `if (protocolValue)`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+    URL dummyURL("https://www.webkit.org"_s);
+
+    if (protocolValue.has_value())
+        dummyURL.setProtocol(protocolValue.value());
```
No big deal but this is a bit shorter `*protocolValue` and you have used it earlier in your PR.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+        || protocolValue == "http"_s || protocolValue == "https"_s
+        || protocolValue == "ws"_s || protocolValue == "wss"_s) {
+
+        bool hasLeadingSlash = pathnameValue.startsWith('/');
```
Could use `pathnameValue[0] == '/'` since you've checked `pathnameValue.isEmpty()` above.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+        bool hasLeadingSlash = pathnameValue.startsWith('/');
+        auto modifiedValue = hasLeadingSlash ? ""_s : "/-"_s;
+
+        auto maybeAddSlashPrefix = makeString(modifiedValue, pathnameValue);
```
Seems this would be a bit more efficient:
```
auto maybeAddSlashPrefix = hasLeadingSlash ? pathnameValue : makeString("/-"_s, pathnameValue);
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+        if (!dummyURL.isValid())
+            return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL path string."_s };
+
+        auto result = dummyURL.path().toString();
```
Very sad for perf to convert to a String just to be calling substring below and re-allocating yet another string. This would be more efficient:
```
auto result = dummyURL.path();
if (hasLeadingSlash)
    result = result.substring(2);
return result.toString();
```

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+    if (value.isEmpty())
+        return String { value };
+
+    auto strippedValue = value.startsWith('?') ? value.substring(1) : value;
```
Could use `value[0] == '?'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+    if (value.isEmpty())
+        return String { value };
+
+    auto strippedValue = value.startsWith('#') ? value.substring(1) : value;
```
could use `value[0] == '#'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-port, combined with https://urlpattern.spec.whatwg.org/#process-port-for-init
+ExceptionOr<String> canonicalizePort(const String& portValue, std::optional<StringView> protocolValue, BaseURLStringType portValueType)
```
Why is this taking a `std::optional<StringView>` for protocolValue instead of a `StringView`?

jelee53 replied:
spec screenshot: https://urlpattern.spec.whatwg.org/#canonicalize-a-port
<img width="575" alt="Screenshot 2024-10-21 at 4 57 54 PM" src="https://github.com/user-attachments/assets/e2d290df-ad35-4516-a5b3-13e73c6c8906">

jelee53 replied:
The spec wanted it to be optional.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    } else if (std::holds_alternative<URLPatternInit>(input.value()))
+        init = std::get<URLPatternInit>(input.value());
+
+    auto maybeProcessedInit = processInit(WTFMove(init), BaseURLStringType::Pattern, String(), String(), String(), String(), String(), String(), String(), String());
```
Is passing `String()` like this for most parameters a temporary thing due to incomplete implementation or how it's going to be longer term? If that how it's going to be longer term, I would add default parameter values on the function so that the caller can simply omit those parameters. If it's temporary, I would add a FIXME comment.

jelee53 replied:
I'll add default parameters. I don't think the processInit is called anywhere but during the "create" function and "create" specifies that these fields should always be null.

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    URLPatternInit init;
+
+    if (std::holds_alternative<String>(input.value()) && !std::get<String>(input.value()).isNull()) {
+        // FIXME: Implement converting String input to URLPatternInit.
```
Might be better return Exception if we don't intend to implement it in this patch. It's easier for tests to catch Exception than unexpected behavior due to missing implementation.

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    }
+
+    if (!init.protocol.isNull()) {
+        auto protocolResult = canonicalizeProtocol(init.protocol, type);
```
It looks like all these canonicalize* functions need `dummyURL` and we are creating it for each function. Can we just create one `dummyURL` and passes its reference to these functions? This will help avoid parsing on URL creation.

jelee53 replied:
Chris and Ryosuke discussed it earlier and they said that they wanted to leave as-is for now. 

<img width="836" alt="Screenshot 2024-10-22 at 11 55 20 AM" src="https://github.com/user-attachments/assets/9e06cf35-22a4-452b-818c-914fd52ffee4">

cdumez replied:
I did suggest using a global static constant for the ASCIILiteral URL though, for avoid duplication. We can optimize this part later on I think.

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,247 @@
+    result.port = port;
+    result.pathname = pathname;
+    result.search = search;
+    result.hash = hash;
```
`result.baseURL` is not set

jelee53 replied:
Oh! For this one, the spec said that it only wanted these specific fields set: 
<img width="434" alt="Screenshot 2024-10-22 at 11 36 26 AM" src="https://github.com/user-attachments/assets/1a9ac94f-fe99-4c5c-b01c-2bb33d5aed8b">

szewai replied:
ah I see, so  `baseURL` is not inherited

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,245 @@
-URLPattern::URLPattern() = default;
+URLPattern::URLPattern(URLPatternInit&& initInput)
+{
+    m_protocol = initInput.protocol;
```
These should be initialized via the initialization list.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,29 +26,245 @@
+    m_hash = initInput.hash;
+}
 
+URLPattern::URLPattern() = default;
```
Do we still need this one?

jelee53 replied:
The Blink implementation has it so I'd like to keep it around.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.h:
```diff
@@ -62,6 +63,7 @@ class URLPattern final : public RefCounted<URLPattern> {
     bool hasRegExpGroups() const { return m_hasRegExpGroups; }
 
 private:
+    URLPattern(URLPatternInit&& initInput);
```
`explicit`

cdumez reviewed (approved):
LGTM with minor fixes.

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,28 +26,243 @@
+}
+
+// https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit
+static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStringType type, String&& protocol = String(), String&& username = String(), String&& password = String(), String&& hostname = String(), String&& port = String(), String&& pathname = String(), String&& search = String(), String&& hash = String())
```
Should be bracket: `String&& protocol = { }`

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,28 +26,243 @@
+    if  (!init.baseURL.isNull()) {
+        baseURL = URL(init.baseURL);
+
+        if (!baseURL.isValid() || baseURL.isEmpty())
```
Are we sure empty baseURL is not allowed? Do we have a test for this?

cdumez replied:
> Do we have a test for this?

Tests are not enabled yet

jelee53 replied:
The subsequent code is using baseURL.user(), baseURL.password(), etc to populate the URLPattern. 

URLParser accepts the empty string as valid input so I guess I thought it made sense because if the baseURL string is empty, these fields should be null-opt anyways. Which is no change from the initial initializations. 

eg - port() is not going to be different (aka is null anyways) if we process an empty baseURL string. 
<img width="786" alt="Screenshot 2024-10-23 at 10 09 05 AM" src="https://github.com/user-attachments/assets/30c2bc01-d42e-4faf-968b-f852bbbd6f13">

szewai replied:
I understand what you said, but you are throwing exception on empty URL here, which means it is an invalid parameter. From your description, it's a valid parameter, it's just we are not going to update some fields of result based on `baseURL`, right?

jelee53 replied:
I guess my thought was it's better to throw a failure rather than to have some side-effects like updating Protocol to an empty String instead of a default null String, etc.

szewai reviewed (commented):
Can you update the URLPatternAPIEnabled flag to `testable` and update expectations for imported tests in this patch?
It seems hard to validate the implementation of algorithms without any test.

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,28 +26,243 @@
+// https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit
+static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStringType type, String&& protocol = String(), String&& username = String(), String&& password = String(), String&& hostname = String(), String&& port = String(), String&& pathname = String(), String&& search = String(), String&& hash = String())
+{
+    URLPatternInit result;
```
We can directly construct result with the parameters (`baseURL` can be nullString, i.e. `{ }`):
```
URLPatternInit result {
    WTFMove(protocol);
    WTFMove(username);
    ...
}
```

szewai commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -26,28 +26,243 @@
+    , m_port(initInput.port)
+    , m_pathname(initInput.pathname)
+    , m_search(initInput.search)
+    , m_hash(initInput.hash)
```
These should use `WTFMove`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
```diff
@@ -0,0 +1,257 @@
+    if (valueType == BaseURLStringType::Pattern)
+        return strippedValue;
+
+    URL dummyURL(makeString(strippedValue, "://webkit.test"_s));
```
Can we just use `dummyURLCharacters` for consistency?

jelee53 replied:
I need to append stuff that makes the protocol string look like a protocol. 

I couldn't do https://www.webkit.org because it already as a protocol. Also the spec makes a distinction between canonicalizing protocol and creating dummy URL's for the other, more generic fields.

 https://urlpattern.spec.whatwg.org/#canonicalize-a-protocol
<img width="943" alt="Screenshot 2024-10-23 at 11 29 40 AM" src="https://github.com/user-attachments/assets/67b3d532-6df4-4882-9865-681412b196e9">

jelee53 committed:
Implement URLPatternInit support for URLPatternAPI.
https://bugs.webkit.org/show_bug.cgi?id=281711
rdar://138169179

Reviewed by Chris Dumez and Sihui Liu.

URLPattern API spec for processing a URLPatternInit: https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::escapePatternStringForCharacters):
(WebCore::escapePatternString):
(WebCore::processBaseURLString):
(WebCore::processInit):
(WebCore::URLPattern::create):
(WebCore::URLPattern::URLPattern):
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp: Added.
(WebCore::isInvalidIPv6HostCodePoint):
(WebCore::isAbsolutePathname):
(WebCore::canonicalizeProtocol):
(WebCore::canonicalizeUsername):
(WebCore::canonicalizePassword):
(WebCore::canonicalizeHost):
(WebCore::canonicalizeIPv6Host):
(WebCore::canonicalizePort):
(WebCore::canonicalizeOpaquePath):
(WebCore::canonicalizePath):
(WebCore::canonicalizeSearch):
(WebCore::canonicalizeHash):
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.h: Copied from Source/WebCore/Modules/url-pattern/URLPattern.cpp.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/285640@main
